File: packages/protocol-ts/src/envelope.ts
/**
 * TypeScript mirror of the Python3 protocol v1 envelope and replay cursors.
 * Python3 reference: src/l2/protocol/envelope.py — keep in sync (§2.4).
 *
 * Re-exports from types.ts (branded IDs + discriminated payloads) plus the
 * ring-buffer Outbox. Consumers import from this module as the entry point.
 */

// ── Type re-exports from types.ts ──────────────────────────────────────
export {
  PROTOCOL_VERSION, OUTBOX_MAXLEN, KINDS, CONTROL_OPS,
  asSessionId, asViewId,
} from "./types.ts";
export type {
  MessageKind, ControlOp,
  SessionId, ViewId,
  AckPayload, CommandPayload, ControlPayload,
  EventPayload, IntentPayload, ResultPayload, StreamChunkPayload,
  TypedMessage, DecodedMessage, Message,
} from "./types.ts";
export type { JsonObject, JsonValue } from "./records.ts";

import type { JsonObject } from "./records.ts";
import type { Message, MessageKind, DecodedMessage } from "./types.ts";
import { canonicalJson } from "./records.ts";
import { PROTOCOL_VERSION, OUTBOX_MAXLEN } from "./types.ts";

// ── Runtime helpers ────────────────────────────────────────────────────

function isObject(value: unknown): value is JsonObject {
  return typeof value === "object" && value !== null && !Array.isArray(value);
}

function isInteger(value: unknown, minimum = 0): value is number {
  return typeof value === "number" && Number.isInteger(value) && value >= minimum;
}

function nowSeconds(): number {
  return Date.now() / 1000;
}

/** Build a protocol v1 message with the same defaults as Python3. */
export function makeMessage(
  sessionId: string,
  seq: number,
  kind: MessageKind,
  payload: JsonObject,
  traceId = "",
  ts = nowSeconds(),
): Message {
  return { v: PROTOCOL_VERSION, session_id: sessionId, seq, ts, trace_id: traceId, kind, payload };
}

// ── Payload validation ─────────────────────────────────────────────────

const CONTROL_OPS_SET = new Set<string>(["attach", "detach", "resume", "recovery", "ack"]);

function validatePayload(kind: MessageKind, payload: JsonObject): string[] {
  const errors: string[] = [];
  if (kind === "command") {
    if (typeof payload.name !== "string" || payload.name.length === 0)
      errors.push("command payload requires a non-empty name");
    if (payload.args !== undefined && (!Array.isArray(payload.args) || payload.args.some((arg) => typeof arg !== "string")))
      errors.push("command payload args must be a string array");
  } else if (kind === "intent") {
    if (typeof payload.text !== "string" || payload.text.length === 0)
      errors.push("intent payload requires non-empty text");
  } else if (kind === "result") {
    if (typeof payload.success !== "boolean")
      errors.push("result payload requires boolean success");
  } else if (kind === "stream_chunk") {
    if (typeof payload.data !== "string")
      errors.push("stream_chunk payload requires string data");
  } else if (kind === "control") {
    const op = payload.op;
    if (typeof op !== "string" || !CONTROL_OPS_SET.has(op))
      errors.push(`control payload has unknown op: ${String(op)}`);
    if (payload.session_id !== undefined && (typeof payload.session_id !== "string" || (payload.session_id as string).length === 0))
      errors.push("control payload session_id must be a non-empty string");
    if (payload.last_acked !== undefined && !isInteger(payload.last_acked, -1))
      errors.push("control payload last_acked must be an integer >= -1");
  } else if (kind === "ack" && !isInteger(payload.ack_seq)) {
    errors.push("ack payload requires a non-negative integer ack_seq");
  }
  return errors;
}

// ── Envelope validation ────────────────────────────────────────────────

const REQUIRED_FIELDS = new Set(["v", "session_id", "seq", "ts", "kind", "payload"]);
const KINDS_SET = new Set<string>(["ack", "command", "control", "event", "intent", "result", "stream_chunk"]);

export function validateMessage(message: unknown): string[] {
  /** Return validation errors; empty = valid. */
  if (!isObject(message)) return ["envelope must be an object"];
  const rec = message as Record<string, unknown>;
  const errors: string[] = [];
  for (const field of REQUIRED_FIELDS) {
    if (!(field in rec)) errors.push(`missing field: ${field}`);
  }
  if (errors.length > 0) return errors;
  if (rec.v !== PROTOCOL_VERSION) errors.push(`unsupported version: ${String(rec.v)}`);
  if (typeof rec.session_id !== "string" || rec.session_id.length === 0)
    errors.push("session_id must be a non-empty string");
  if (!isInteger(rec.seq)) errors.push("seq must be a non-negative integer");
  if (typeof rec.kind !== "string" || !KINDS_SET.has(rec.kind))
    errors.push(`unknown kind: ${String(rec.kind)}`);
  if (typeof rec.ts !== "number" || !Number.isFinite(rec.ts))
    errors.push("ts must be a number");
  if ("trace_id" in rec && typeof rec.trace_id !== "string")
    errors.push("trace_id must be a string");
  if (!isObject(rec.payload)) {
    errors.push("payload must be an object");
  } else if (typeof rec.kind === "string" && KINDS_SET.has(rec.kind)) {
    errors.push(...validatePayload(rec.kind as MessageKind, rec.payload));
  }
  return errors;
}

export function encodeMessage(message: Message): string {
  /** Serialize a valid envelope with recursively sorted keys. */
  const errors = validateMessage(message);
  if (errors.length > 0) throw new Error(errors.join("; "));
  return canonicalJson(message);
}

export function decodeMessage(line: string): DecodedMessage {
  /** Decode and validate one JSONL envelope without throwing parse errors. */
  if (typeof line !== "string" || line.trim().length === 0) return { message: null, error: "empty line" };
  let raw: unknown;
  try {
    raw = JSON.parse(line);
  } catch (error) {
    return { message: null, error: `invalid json: ${String(error)}` };
  }
  const errors = validateMessage(raw);
  if (errors.length > 0) return { message: null, error: errors.join("; ") };
  return { message: raw as Message, error: null };
}

// ── Outbox (bounded replay window) ───────────────────────────────

/**
 * Bounded per-session replay window (simple array, maxlen 1024).
 *
 * `Array.shift()` is O(n) but n≤1024 so cost is negligible vs. the
 * ring-buffer's extra `head/count` state; kept simple for auditability.
 * Non-destructive `ack` mirrors `src/l2/protocol/envelope.py`.
 */
export class Outbox {
  private items: Message[] = [];
  private acknowledged = -1;

  constructor(public readonly maxlen = OUTBOX_MAXLEN) {
    if (!Number.isInteger(maxlen) || maxlen < 1) throw new Error("maxlen must be a positive integer");
  }

  /** Append, evicting oldest when at capacity. */
  append(message: Message): void {
    this.items.push(message);
    if (this.items.length > this.maxlen) this.items.shift();
  }

  /** Non-destructive ack: only cursor advances. */
  ack(seq: number): void {
    this.acknowledged = Math.max(this.acknowledged, seq);
  }

  /** Replay window after `afterSeq` (oldest-first). */
  unacked(afterSeq?: number): Message[] {
    const after = afterSeq ?? this.acknowledged;
    return this.items.filter((m) => m.seq > after);
  }

  /** Last acknowledged seq. */
  get lastAcked(): number {
    return this.acknowledged;
  }

  /** Buffered count (≤ maxlen). */
  get size(): number {
    return this.items.length;
  }
}

// ── SessionCursor (per-view ack cursor) ────────────────────────────────

/**
 * Per-view ack cursor mirroring the Python3 host's view tracking.
 *
 * One cursor per frontend view; host outbox stays authoritative.
 */
export class SessionCursor {
  sessionId = "";
  lastAcked = -1;
  attached = false;

  constructor(public readonly viewId: string) {}

  /** Bind the view to a session. */
  attach(sessionId: string): void {
    this.sessionId = sessionId;
    this.attached = true;
  }

  /** Detach the view (no longer receives replay). */
  detach(): void {
    this.attached = false;
  }

  /** Advance the acknowledged position (monotonic). */
  ack(seq: number): void {
    this.lastAcked = Math.max(this.lastAcked, seq);
  }
}

File: src/l2/protocol/envelope.py
"""Protocol v1 envelope — frozen legacy reference (side-effect free).

DEPRECATED as normative source: `packages/protocol-ts` is the authority
(docs/architecture/l2-shell-engine.md, "Protocol v1 conformance rulings").
This module is kept byte-compatible with the TS/Rust semantics until the
G6 cut-over retires it: non-destructive `Outbox.ack` (R1), finite-`ts`
validation and encoding (R3). Stdlib-only, no singletons, no I/O.
"""

from __future__ import annotations

import json
import math
import time
from collections import deque
from dataclasses import dataclass, field
from typing import Any

PROTOCOL_VERSION: int = 1

KIND_INTENT = "intent"
KIND_COMMAND = "command"
KIND_EVENT = "event"
KIND_RESULT = "result"
KIND_STREAM_CHUNK = "stream_chunk"
KIND_CONTROL = "control"
KIND_ACK = "ack"
KINDS: frozenset[str] = frozenset(
    {KIND_INTENT, KIND_COMMAND, KIND_EVENT, KIND_RESULT, KIND_STREAM_CHUNK, KIND_CONTROL, KIND_ACK}
)

ENVELOPE_FIELDS: tuple[str, ...] = ("v", "session_id", "seq", "ts", "kind", "payload")

# Bounded replay window per session (recovery reads this, never the past).
OUTBOX_MAXLEN: int = 1024

CONTROL_ATTACH = "attach"
CONTROL_DETACH = "detach"
CONTROL_RESUME = "resume"
CONTROL_RECOVERY = "recovery"
CONTROL_ACK = "ack"
CONTROL_KINDS: frozenset[str] = frozenset(
    {CONTROL_ATTACH, CONTROL_DETACH, CONTROL_RESUME, CONTROL_RECOVERY, CONTROL_ACK}
)

# R4: authorization fields are host-derived (adapter-injected GateRequest
# inputs) and must never travel on the wire.
HOST_DERIVED_FIELDS: tuple[str, ...] = (
    "approved",
    "pre_approved",
    "full_power",
    "harness_auto_approved",
)


def make_message(
    session_id: str,
    seq: int,
    kind: str,
    payload: dict[str, Any],
    *,
    trace_id: str = "",
    ts: float | None = None,
) -> dict[str, Any]:
    """Build a protocol v1 envelope dict from parts."""
    return {
        "v": PROTOCOL_VERSION,
        "session_id": session_id,
        "seq": seq,
        "ts": time.time() if ts is None else ts,
        "trace_id": trace_id,
        "kind": kind,
        "payload": payload,
    }


def validate_message(msg: dict[str, Any]) -> list[str]:
    """Return contract violations as a list; an empty list means valid."""
    errors: list[str] = []
    for field_name in ENVELOPE_FIELDS:
        if field_name not in msg:
            errors.append(f"missing field: {field_name}")
    if errors:
        return errors
    if msg.get("v") != PROTOCOL_VERSION:
        errors.append(f"unsupported version: {msg.get('v')!r}")
    session_id = msg.get("session_id")
    if not isinstance(session_id, str) or not session_id:
        errors.append("session_id must be a non-empty string")
    seq = msg.get("seq")
    if isinstance(seq, bool) or not isinstance(seq, int) or seq < 0:
        errors.append("seq must be a non-negative integer")
    kind = msg.get("kind")
    if not isinstance(kind, str) or kind not in KINDS:
        errors.append(f"unknown kind: {kind!r}")
    ts = msg.get("ts")
    if isinstance(ts, bool) or not isinstance(ts, (int, float)) or not math.isfinite(ts):
        errors.append("ts must be a number")
    if "trace_id" in msg and not isinstance(msg["trace_id"], str):
        errors.append("trace_id must be a string")
    payload = msg.get("payload")
    if not isinstance(payload, dict):
        errors.append("payload must be an object")
    elif isinstance(kind, str) and kind in KINDS:
        errors.extend(_validate_payload(kind, payload))
    return errors


def _validate_payload(kind: str, payload: dict[str, Any]) -> list[str]:
    """Validate the required fields for one message kind."""
    errors: list[str] = []
    if kind in (KIND_COMMAND, KIND_CONTROL) and any(f in payload for f in HOST_DERIVED_FIELDS):
        errors.append("payload carries host-derived authorization fields")
    if kind == KIND_COMMAND:
        name = payload.get("name")
        args = payload.get("args", [])
        if not isinstance(name, str) or not name:
            errors.append("command payload requires a non-empty name")
        if not isinstance(args, list) or any(not isinstance(arg, str) for arg in args):
            errors.append("command payload args must be a string array")
    elif kind == KIND_INTENT:
        if not isinstance(payload.get("text"), str) or not payload["text"]:
            errors.append("intent payload requires non-empty text")
    elif kind == KIND_RESULT:
        if not isinstance(payload.get("success"), bool):
            errors.append("result payload requires boolean success")
    elif kind == KIND_STREAM_CHUNK:
        if not isinstance(payload.get("data"), str):
            errors.append("stream_chunk payload requires string data")
    elif kind == KIND_CONTROL:
        op = payload.get("op")
        if not isinstance(op, str) or op not in CONTROL_KINDS:
            errors.append(f"control payload has unknown op: {op!r}")
        target = payload.get("session_id")
        if target is not None and (not isinstance(target, str) or not target):
            errors.append("control payload session_id must be a non-empty string")
        last_acked = payload.get("last_acked")
        if last_acked is not None and (
            isinstance(last_acked, bool) or not isinstance(last_acked, int) or last_acked < -1
        ):
            errors.append("control payload last_acked must be an integer >= -1")
    elif kind == KIND_ACK:
        ack_seq = payload.get("ack_seq")
        if isinstance(ack_seq, bool) or not isinstance(ack_seq, int) or ack_seq < 0:
            errors.append("ack payload requires a non-negative integer ack_seq")
    return errors


def encode_message(msg: dict[str, Any]) -> str:
    """Serialize one envelope to a canonical JSON line (stable key order).

    Non-finite floats (NaN/Infinity) raise ValueError: R3 forbids emitting
    frames that strict JSON parsers on the Rust/TS side cannot read.
    """
    return json.dumps(msg, ensure_ascii=False, allow_nan=False, sort_keys=True, separators=(",", ":"))


def decode_message(line: str) -> tuple[dict[str, Any] | None, str | None]:
    """Parse one JSON line into a validated envelope; returns (msg, error)."""
    line = line.strip()
    if not line:
        return None, "empty line"
    try:
        msg = json.loads(line)
    except json.JSONDecodeError as e:
        return None, f"invalid json: {e}"
    if not isinstance(msg, dict):
        return None, "envelope must be a JSON object"
    errors = validate_message(msg)
    if errors:
        return None, "; ".join(errors)
    return msg, None


@dataclass
class Outbox:
    """Bounded per-session replay window; unacked messages survive recovery.

    R1 (non-destructive ack): one view's acknowledgement only advances the
    cursor and never erases retained messages another view still needs.
    """

    maxlen: int = OUTBOX_MAXLEN
    _items: deque[dict[str, Any]] = field(default_factory=deque)
    _last_acked: int = -1

    def append(self, msg: dict[str, Any]) -> None:
        """Append an outbound message; evict oldest beyond the cap."""
        self._items.append(msg)
        while len(self._items) > self.maxlen:
            self._items.popleft()

    def ack(self, seq: int) -> None:
        """Advance the acknowledged cursor without dropping retained messages."""
        self._last_acked = max(self._last_acked, seq)

    def unacked(self, after_seq: int | None = None) -> list[dict[str, Any]]:
        """Return the replay window for one view cursor (messages after it)."""
        after = self._last_acked if after_seq is None else after_seq
        return [msg for msg in self._items if msg["seq"] > after]

    @property
    def last_acked(self) -> int:
        """Return the highest acknowledged sequence id."""
        return self._last_acked


@dataclass
class SessionCursor:
    """Per-frontend-view cursor: attachment + acknowledged position."""

    view_id: str
    session_id: str = ""
    last_acked: int = -1
    attached: bool = False

    def attach(self, session_id: str) -> None:
        """Bind the view to a session and mark it attached."""
        self.session_id = session_id
        self.attached = True

    def detach(self) -> None:
        """Unbind the view from its session (marks it unattached)."""
        self.attached = False
